Index: config/minimal.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/config/minimal.toml b/config/minimal.toml
--- a/config/minimal.toml	(revision b82bffcab9d7b9603cdaa213b98ea2e0a3a172f9)
+++ b/config/minimal.toml	(revision 27070d834fa4d80ff49a03e5878372ae28681f14)
@@ -13,6 +13,7 @@
 extrautils = {}
 ion = {}
 smith = {}
+lmbench = {}

 [[files]]
 path = "/usr/lib/init.d/30_console"
Index: mk/qemu.mk
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mk/qemu.mk b/mk/qemu.mk
--- a/mk/qemu.mk	(revision b82bffcab9d7b9603cdaa213b98ea2e0a3a172f9)
+++ b/mk/qemu.mk	(revision 27070d834fa4d80ff49a03e5878372ae28681f14)
@@ -112,7 +112,7 @@
 		QEMUFLAGS+=-device qemu-xhci -device usb-kbd -device usb-tablet
 	endif
 else
-$(error Unsupported ARCH for QEMU "$(ARCH)"))
+$(error Unsupported ARCH for QEMU "$(ARCH)")
 endif

 QEMUFLAGS+=-smp $(QEMU_SMP) -m $(QEMU_MEM)
@@ -300,5 +300,7 @@
 #additional steps for $(DISK) are required!!!
 qemu_raspi: qemu-deps
 	$(QEMU) -M raspi3b -smp 4,cores=1 \
-		-kernel $(FIRMWARE) \
+		-kernel $(QEMU_KERNEL) \
+		-sd $(DISK) \
 		-serial stdio -display none
