diff --git a/storage/bcm2835-sdhcid/src/main.rs b/storage/bcm2835-sdhcid/src/main.rs
index 878604a..b47584f 100644
--- a/storage/bcm2835-sdhcid/src/main.rs
+++ b/storage/bcm2835-sdhcid/src/main.rs
@@ -73,7 +73,9 @@ fn daemon(daemon: redox_daemon::Daemon) -> ! {
     );
     let mut sdhci = sd::SdHostCtrl::new(addr);
     unsafe {
-        sdhci.init();
+        // sdhci.init();
+        sdhci.size=0x76e480000;//this should be set to the size of your sd-card
+        sdhci.scr[0] = 0x0200_0001;
         /*
         let mut buf1 = [0u32; 512];
         sdhci.sd_readblock(1, &mut buf1, 1);
diff --git a/storage/bcm2835-sdhcid/src/sd/mod.rs b/storage/bcm2835-sdhcid/src/sd/mod.rs
index 00d9677..52ce19c 100644
--- a/storage/bcm2835-sdhcid/src/sd/mod.rs
+++ b/storage/bcm2835-sdhcid/src/sd/mod.rs
@@ -174,9 +174,9 @@ pub struct SdHostCtrl {
     cid: [u32; 4],
     csd: [u32; 4],
     rca: u32, //relative card address
-    scr: [u32; 2],
+    pub scr: [u32; 2],
     ocr: u32,
-    size: u64,
+    pub size: u64,
 }
 
 impl SdHostCtrl {
